akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -149,7 +149,21 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
   private def updateState(streamId: Int, handle: StreamState => StreamState, event: String, eventArg: AnyRef = null): Unit =
     updateStateAndReturn(streamId, x => (handle(x), ()), event, eventArg)
 
+  // Calling multiplexer.enqueueOutStream directly out of the state machine is not allowed, because it might try to
+  // reenter the state machine with `pullNextState`. This call defers enqueuing until the current state machine operation
+  // is done.
+  private var deferredStreamToEnqueue: Int = -1
+  private def enqueueOutStream(streamId: Int): Unit = if (stateMachineRunning) {
+    require(deferredStreamToEnqueue == -1, "Only one stream can be enqueued during a single state change")
+    deferredStreamToEnqueue = streamId
+  } else
+    multiplexer.enqueueOutStream(streamId)
+
+  private var stateMachineRunning = false
   private def updateStateAndReturn[R](streamId: Int, handle: StreamState => (StreamState, R), event: String, eventArg: AnyRef = null): R = {
+    require(!stateMachineRunning, "State machine already running")
```

Comment:
Is there a possibility that calling `tryPullSubStreams()` below leads to reentering `updateStateAndReturn` while `stateMachineRunning` is still `true`?

Reply:
I don't think so, `tryPullSubStreams` only calls methods of the streams infrastructure which will not execute directly on the stack.